extensions/cornerstone/src/services/CornerstoneCacheService/CornerstoneCacheService.ts

```diff
@@ -0,0 +1,252 @@
+import { cache as cs3DCache, Enums, volumeLoader } from '@cornerstonejs/core';
+import { utils } from '@ohif/core';
+
+import getCornerstoneViewportType from '../../utils/getCornerstoneViewportType';
+import {
+  StackViewportData,
+  VolumeViewportData,
+} from '../../types/CornerstoneCacheService';
+
+const VOLUME_IMAGE_LOADER_SCHEME = 'streaming-wadors';
+const VOLUME_LOADER_SCHEME = 'cornerstoneStreamingImageVolume';
+
+class CornerstoneCacheService {
+  stackImageIds: Map<string, string[]> = new Map();
+  volumeImageIds: Map<string, string[]> = new Map();
+
+  constructor(servicesManager) {
+    this.servicesManager = servicesManager;
+  }
+
+  public getCacheSize() {
+    return cs3DCache.getCacheSize();
+  }
+
+  public getCacheFreeSpace() {
+    return cs3DCache.getBytesAvailable();
+  }
+
+  public async createViewportData(
+    displaySets: unknown[],
+    viewportOptions: Record<string, unknown>,
+    dataSource: unknown,
+    initialImageIndex?: number
+  ): Promise<StackViewportData | VolumeViewportData> {
+    let viewportType = viewportOptions.viewportType as string;
+
+    // Todo: Since Cornerstone 3D currently doesn't support segmentation
+    // on stack viewport, we should check if whether the the displaySets
+    // that are about to be displayed are referenced in a segmentation
+    // as a reference volume, if so, we should hang a volume viewport
+    // instead of a stack viewport
+    if (this._shouldRenderSegmentation(displaySets)) {
+      viewportType = 'volume';
+
+      // update viewportOptions to reflect the new viewport type
+      viewportOptions.viewportType = viewportType;
+    }
+
+    const cs3DViewportType = getCornerstoneViewportType(viewportType);
+    let viewportData: StackViewportData | VolumeViewportData;
+
+    if (cs3DViewportType === Enums.ViewportType.STACK) {
+      viewportData = await this._getStackViewportData(
+        dataSource,
+        displaySets,
+        initialImageIndex
+      );
+    }
+
+    if (cs3DViewportType === Enums.ViewportType.ORTHOGRAPHIC) {
+      viewportData = await this._getVolumeViewportData(dataSource, displaySets);
+    }
+
+    viewportData.viewportType = cs3DViewportType;
+
+    return viewportData;
+  }
+
+  public async invalidateViewportData(
```

Comment:
What is this used for? Not sure I get it

Reply:
When displaySet gets invalidated (patient weight change)